Support qcow2 external data file

XueqiangWei opened the pull request:
Add support for qcow2 external data file.
New parameters:

`enable_data_file`: works when `image_format` is qcow2, will create
external data file(`root_dir/images/{image}.data_file` or specified by
`image_data_file_path`).

`image_data_file_path`: optional, used to specify the path to external
data file.

`image_data_file_raw`: if the data file is kept as a raw file.

id: 1828649
Signed-off-by: Xueqiang Wei <xuwei@redhat.com>

XueqiangWei commented:
add enable_data_file = yes to unattended_install.cfg, and then run it. It works well.

(1/3) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1973.54 s)
(2/3) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.rh_kernel_update.q35: PASS (137.80 s)
(3/3) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot.q35: PASS (56.36 s)

XueqiangWei commented:
set data_file_path in cfg, it works well.
enable_data_file = yes
data_file_path = /home/test.data_file
data_file_raw = on
(1/1) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot.q35: PASS (55.11 s)

XueqiangWei commented:
@luckyh please help review it. Many thanks.

xutian commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -2224,7 +2242,10 @@ def images_define_by_params(self, name, image_params, media=None,
                                                image_encryption,
-                                               image_access)
+                                               image_access,
+                                               (ext_data_file and
```
pass ext_data_file  directly.

luckyh replied:
Agreed.

XueqiangWei replied:
updated, thanks.

xutian commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1926,6 +1928,18 @@ def define_hbas(qtype, atype, bus, unit, port, qbus, pci_bus, iothread,
                 elif imgfmt == "luks":
                     devices[-1].set_param('key-secret', secret_obj.get_qid())
 
+        if external_data_file:
```
```
extra_date_file = getatrr(image_filename, "image_filename", None)
if extra_date_file:
     ...
```

XueqiangWei replied:
updated, thanks.

xutian commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1550,7 +1551,7 @@ def images_define_by_variables(self, name, filename, pci_bus, index=None,
                                    num_queues=None, bus_extra_params=None,
                                    force_fmt=None, image_encryption=None,
-                                   image_access=None):
+                                   image_access=None, external_data_file=None):
```
let's pass image object directly.

luckyh commented on virttest/storage.py:
```diff
@@ -618,7 +619,31 @@ def __init__(self, params, root_dir, tag):
+        enable_data_file = params.get("enable_data_file", "no") == "yes"
+        image_format = params.get("image_format", "qcow2")
+        if image_format != "qcow2" or not enable_data_file:
+            return
```
Should raise `ValueError` for those formats don't support external data file.

XueqiangWei replied:
update, thanks

XueqiangWei commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1926,6 +1928,18 @@ def define_hbas(qtype, atype, bus, unit, port, qbus, pci_bus, iothread,
+                if stat.S_ISBLK(mode):
+                    ext_driver = "host_device"
+            devices[-1].set_param('data-file.driver', ext_driver)
+            devices[-1].set_param('data-file.filename', ext_filename)
```
yes, and I tested with general file and host block device, all work well. The command lines as below.

**general file + blockdev:**
-blockdev node-name=file_image1,driver=file,aio=threads,filename=/home/kvm_autotest_root/images/rhel820-64-virtio-scsi.qcow2,cache.direct=on,cache.no-flush=off \
-blockdev node-name=drive_image1,driver=qcow2,data-file.driver=file,data-file.filename=/var/lib/avocado/data/avocado-vt/images/rhel820-64-virtio-scsi.data_file,cache.direct=on,cache.no-flush=off,file=file_image1 \

**general file + drive:**
-drive id=drive_image1,if=none,snapshot=off,data-file.driver=file,data-file.filename=/var/lib/avocado/data/avocado-vt/images/rhel820-64-virtio-scsi.data_file,aio=threads,cache=none,format=qcow2,file=/home/kvm_autotest_root/images/rhel820-64-virtio-scsi.qcow2 \

**block device + blockdev:**
-blockdev node-name=file_image1,driver=file,aio=threads,filename=/home/kvm_autotest_root/images/rhel820-64-virtio-scsi.qcow2,cache.direct=on,cache.no-flush=off \
-blockdev node-name=drive_image1,driver=qcow2,data-file.driver=host_device,data-file.filename=/dev/mapper/mpatha,cache.direct=on,cache.no-flush=off,file=file_image1 \

**block device + drive:**
-drive id=drive_image1,if=none,snapshot=off,data-file.driver=host_device,data-file.filename=/dev/mapper/mpatha,aio=threads,cache=none,format=qcow2,file=/home/kvm_autotest_root/images/rhel820-64-virtio-scsi.qcow2 \

XueqiangWei commented:
> Also, seems missing the construction/destruction procedure for external data files in `preprocess_image`/`postprocess_image`, would you please clarify that? thanks.
Class QemuImg will be invoked in `preprocess_image`/`postprocess_image`.
I added related methods to QemuImg.

XueqiangWei commented:
I found the code conflicts, rebased the codes and updated them.

Tested the latest codes, the result is passed. 
(1/1) Host_RHEL.m8.u3.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot.q35: PASS (73.08 s)

XueqiangWei commented:
@luckyh Please help review it. Many thanks.

CongLi commented:
@luckyh  Could you please help review the patch again ? Thanks.

luckyh commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1994,6 +1996,19 @@ def define_hbas(qtype, atype, bus, unit, port, qbus, pci_bus, iothread,
+        external_data_file = getattr(external_data_file, "image_filename", None)
+        if external_data_file:
+            parse_out = urllib.parse.urlparse(external_data_file, scheme="file")
+            ext_driver, ext_filename = parse_out.scheme, parse_out.path
```
Suggest:
```python
data_file_path = getattr(...)
if data_file_path:
    # by now we only support local files
    data_file_driver = "file"

    if data_file_driver == "file":
        data_file_mode = os.stat(data_file_path).st_mode
        if stat.S_ISBLK(data_file_mode):
           data_file_driver  = "host_device"
    ...
```

XueqiangWei replied:
updated, thanks.

luckyh commented on virttest/storage.py:
```diff
@@ -656,7 +657,34 @@ def __init__(self, params, root_dir, tag):
+        elif image_format != "qcow2":
+            raise ValueError("The %s format does not support external "
+                             "data file" % image_format)
+        else:
```
Should be:
```python
if not enable_data_file:
    return
if image_format != "qcow2":
    raise ValueError(...)
image_size = params["image_size"]
...
```

XueqiangWei replied:
updated, thanks

luckyh commented on virttest/storage.py:
```diff
@@ -656,7 +657,34 @@ def __init__(self, params, root_dir, tag):
+        else:
+            image_size = params["image_size"]
+            base_name = os.path.basename(params["image_name"])
+            data_file_path = params.get("data_file_path",
```
Suggest the name `image_data_file_path`

XueqiangWei replied:
Since the date_file and base image will be created together with a command \"qemu-img create\", I think no need to set parameter to  image_data_file_path. If I was wrong, please correct me. Many thanks.

e.g.
Create a qcow2: external data file image
# qemu-img create -f qcow2 -o data_file=base.data_file,data_file_raw=off base.qcow2 20G

luckyh replied:
Suppose that other type of devices will have the `data_file` option in the future, there could be name conflicts against the params, so that's why I suggest adding the prefix `image_` as its namespace.

XueqiangWei replied:
updated, thanks.

luckyh commented on virttest/storage.py:
```diff
@@ -656,7 +657,34 @@ def __init__(self, params, root_dir, tag):
+            data_file_path = params.get("data_file_path",
+                                        os.path.join(root_dir, "images",
+                                                     "%s.data_file"
+                                                     % base_name))
```
I didn't see any file creation procedure been invoked in `QemuImg.create`, so would you please explain how it is created? thanks.

XueqiangWei replied:
Please refer to the following codes, hope to help you. Thanks.

QemuImg.__init__(self, params, root_dir, tag)
    storage.QemuImg.__init__(self, params, root_dir, tag)

storage.QemuImg.__init__(self, params, root_dir, tag)
    self.data_file = self.external_data_file_defined_by_params(params, root_dir, tag)

QemuImg._parse_options(self, params)
    if self.data_file:
        options.extend(
            ("data_file=%s" % self.data_file.image_filename,
             "data_file_raw=%s" % params.get("data_file_raw", "off")))

QemuImg.create(self, params, ignore_errors=False)
    ......
    options = self._parse_options(params)
    ......

luckyh replied:
I mean, do we need to create the data file explicitly before the qcow2 image creation? In other words, about the management of data file, what actions are required?

I tested the image creation with a non-existent data file, and it turns out that the data file would be created when the `create` command being executed, that means we do can let the data file to be created when it is being used at the first time, but I'm not sure if it is correct for all the usecases. And, for image deletion it seems missing the action for data files, because obviously tools other than `qemu-img` won't remove data files explicitly.

XueqiangWei replied:
for external data file, it's a image file contains only the meta-data and a pointer to the raw device, which could be treated as "thin qcow2".

The date_file and base image will be created together with a command "qemu-img create". 

Customers need a way to track meta-data of raw devices or files, for fetures like incremental backup. So add this function.

Since libvirt won't support external data file, we just cover installation and boot in qemu side. No other operation.

XueqiangWei replied:
As we discussed, added image deletion for data file. thanks.

XueqiangWei commented:
I found the code conflicts, rebased the codes and updated them.
Tested the latest codes, the result is passed.

@luckyh Could you please help review the patch again? Many thanks.

XueqiangWei commented:
Tested without external data file, all work well.
\# python3 ConfigTest.py --testcase=boot --guestname=RHEL.8.3.0 --platform=x86_64 --driveformat=virtio_scsi  --nicmodel=virtio_net --imageformat=qcow2  --machine=q35 --clone=yes

(1/3) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (2431.19 s)
 (2/3) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.rh_kernel_update.q35: PASS (483.92 s)
 (3/3) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.boot.q35: PASS (64.68 s)


Tested with external data file, it works well.
\# python3 ConfigTest.py --testcase=boot --guestname=RHEL.8.2.0 --platform=x86_64 --driveformat=virtio_scsi  --nicmodel=virtio_net --imageformat=qcow2  --machine=q35 --clone=yes --customsparams="enable_data_file = yes"

(1/1) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (3049.12 s)
(1/1) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot.q35: PASS (62.06 s)

Add the below two parameters to unattended_install.cfg, and install a guest with enable_data_file = yes. It works well.
image_data_file_path = /home/external_data_file
image_data_file_raw = on

\# python3 ConfigTest.py --testcase=unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads  --guestname=RHEL.8.3.0 --platform=x86_64 --driveformat=virtio_scsi  --nicmodel=virtio_net --imageformat=qcow2  --machine=q35 --clone=no --customsparams="enable_data_file = yes"

 (1/1) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1899.46 s)

XueqiangWei commented:
@luckyh Please help review it again. Many thanks.

XueqiangWei commented:
@xutian Please help review it. Many thanks.

leidwang commented:
Test this patch with the following command,It works well.
`python3 ConfigTest.py --testcase=boot --guestname=RHEL.8.2.0 --platform=x86_64 --driveformat=virtio_scsi --nicmodel=virtio_net --imageformat=qcow2 --machine=q35 --clone=no --customsparams="enable_data_file = yes"`

 (1/1) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot.q35: PASS (43.87 s)

leidwang commented:
LGTM,
Acked-by: leidwang@redhat.com

luckyh commented:
Hi @xutian, I'd like to merge this PR since all your suggestions have already been applied and there are 2 acks. For any further issues, please feel free to raise it, and we will follow up with another PR accordingly, thanks!

luckyh commented:
Thanks all.
